File: Mathematica-builtin-solver/abcdefghppp.m
#!/usr/local/bin/WolframScript -script

variables = {a, b, c, d, e, f, g, h, p};

solutions = Solve[
    {
        (10 a + b) - (10 c + d) == (10 e + f),  (* AB - CD = EF *)
        (10 e + f) + (10 g + h) == (111*p)      (* EF + GH = PPP *)
    } ~Join~
    (0 < # <= 9 & /@ {a, c, e, g, p}) ~Join~    (* Leading digits are not zero *)
    (0 <= # <= 9 & /@ {b, d, f, h}),            (* The rest can be anything between 0 and 9 *)

    variables,

    Integers        (* Solve in integer domain *)
];

realSolutions = Select[solutions, DuplicateFreeQ[variables /. #]&];  (* all numbers are distinct *)

s = IntegerString;

Print[StringJoin[{s[10a+b]," - ",s[10c+d]," = ",s[10e+f]," + ",s[10g+h]," = ",s[111p]}/.#]]& /@ realSolutions;

(*

Prints:

"85 - 46 = 39 + 72 = 111"
"86 - 54 = 32 + 79 = 111"
"90 - 27 = 63 + 48 = 111"
"90 - 63 = 27 + 84 = 111"
"95 - 27 = 68 + 43 = 111"

*)

File: javascript-1-for-loop-bruteforce/abcdefghppp.js
function test(n){
  var a = n%10; n = Math.floor(n/10);
  var b = n%10; n = Math.floor(n/10);
  var c = n%10; n = Math.floor(n/10);
  var d = n%10; n = Math.floor(n/10);
  var g = n%10; n = Math.floor(n/10);
  var h = n%10; n = Math.floor(n/10); 
  var p = n%10; 
  // constraints
  if( 10*a+b < 10*c+d || 10*a+b-10*c-d+10*g+h != 111*p ) return false;
  var e = Math.floor((10*a+b-10*c-d)/10);
  var f = Math.floor((10*a+b-10*c-d)%10);
  // exclude leading zero
  if( a == 0 || c == 0 || e == 0 ) return false;
  // all digits are distinct
  var arr = [a,b,c,d,e,f,g,h,p];
  for(var i=0; i<arr.length; i++){
    for(var j=i+1; j<arr.length; j++){
      if( arr[i] == arr[j] ) return false;
    }
  }
  return true;
}
function toString(n){
  var a = n%10; n = Math.floor(n/10);
  var b = n%10; n = Math.floor(n/10);
  var c = n%10; n = Math.floor(n/10);
  var d = n%10; n = Math.floor(n/10);
  var e = Math.floor((10*a+b-10*c-d)/10);
  var f = Math.floor((10*a+b-10*c-d)%10);
  var g = n%10; n = Math.floor(n/10);
  var h = n%10; n = Math.floor(n/10); 
  var p = n%10; 
  return [a,b,'-',c,d,'=',e,f].join('') + ' ' + [e,f,'+',g,h,'=',p,p,p].join('');
}
var solution = [];
for(var i=0; i<2e6; i++){
  if( test(i) ) solution.push(i);
}
console.log(solution.map(toString).sort().join('\n'));
console.log('number of solution', solution.length);